crates/pessimistic-proof-utils/src/smt.rs

```diff
@@ -0,0 +1,559 @@
+            return Err(SmtError::KeyPresent);
+        }
+
+        Ok(SmtNonInclusionProof { siblings })
+    }
+}
+
+impl<H, const DEPTH: usize> SmtMerkleProof<H, DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Eq + Hash + Serialize + DeserializeOwned,
+{
+    pub fn verify<K>(&self, key: K, value: H::Digest, root: H::Digest) -> bool
+    where
+        K: ToBits<DEPTH>,
+    {
+        let bits = key.to_bits();
+        let mut hash = value;
+        for i in 0..DEPTH {
+            hash = if bits[DEPTH - i - 1] {
+                H::merge(&self.siblings[i], &hash)
+            } else {
+                H::merge(&hash, &self.siblings[i])
+            };
+        }
+
+        hash == root
+    }
+}
+
+impl<H, const DEPTH: usize> SmtNonInclusionProof<H, DEPTH>
+where
+    H: Hasher,
+    H::Digest: Copy + Eq + Hash + Serialize + DeserializeOwned,
+{
+    pub fn verify<K>(
+        &self,
+        key: K,
+        root: H::Digest,
+        empty_hash_at_height: &[H::Digest; DEPTH],
+    ) -> bool
+    where
+        K: ToBits<DEPTH>,
+    {
+        if self.siblings.len() > DEPTH {
+            return false;
+        }
+        if self.siblings.is_empty() {
+            return root
+                == H::merge(
+                    &empty_hash_at_height[DEPTH - 1],
+                    &empty_hash_at_height[DEPTH - 1],
+                );
+        }
+        let bits = key.to_bits();
+        let mut entry = empty_hash_at_height[DEPTH - self.siblings.len()];
+        for i in (0..self.siblings.len()).rev() {
+            let sibling = self.siblings[i];
+            entry = if bits[i] {
+                H::merge(&sibling, &entry)
+            } else {
+                H::merge(&entry, &sibling)
+            };
+        }
+
+        entry == root
+    }
+
+    /// Verify the non-inclusion proof (i.e. that `key` is not in the SMT) and
+    /// return the updated root of the SMT with `(key, value)` inserted.
+    pub fn verify_and_update<K>(
+        &self,
+        key: K,
+        value: H::Digest,
+        root: H::Digest,
+        empty_hash_at_height: &[H::Digest; DEPTH],
+    ) -> Option<H::Digest>
+    where
+        K: Copy + ToBits<DEPTH>,
+    {
+        if !self.verify(key, root, empty_hash_at_height) {
+            return None;
+        }
+
+        let mut entry = value;
+        let bits = key.to_bits();
+        for i in (self.siblings.len()..DEPTH).rev() {
+            let sibling = empty_hash_at_height[DEPTH - i - 1];
+            entry = if bits[i] {
+                H::merge(&sibling, &entry)
+            } else {
+                H::merge(&entry, &sibling)
+            };
+        }
+        for i in (0..self.siblings.len()).rev() {
+            let sibling = self.siblings[i];
+            entry = if bits[i] {
+                H::merge(&sibling, &entry)
+            } else {
+                H::merge(&entry, &sibling)
+            };
+        }
+
+        Some(entry)
+    }
+}
+
+impl ToBits<32> for u32 {
```

Comment:
Just for my understanding, why implementing it just for 32?

Reply:
Just to give an example for this PR. In #168 it's implemented for more interesting structs.